calculator_test.go

```diff
@@ -2,23 +2,302 @@ package calculator_test
 
 import (
 	"calculator"
+	"math/rand"
 	"testing"
+	"time"
 )
 
 func TestAdd(t *testing.T) {
 	t.Parallel()
-	var want float64 = 4
-	got := calculator.Add(2, 2)
-	if want != got {
-		t.Errorf("want %f, got %f", want, got)
+	testCases := []struct {
+		name  string
+		a, b  float64
+		extra []float64
+		want  float64
+	}{
+		{
+			name:  "Sum of 2 operands to give a positive number",
+			a:     1,
+			b:     1,
+			extra: nil,
+			want:  2,
+		},
+		{
+			name:  "Sum of 3 positive numbers to give a positive number",
+			a:     1,
+			b:     1,
+			extra: []float64{1},
+			want:  3,
+		},
+		{
+			name:  "Sum of 5 negative numbers to give a negative number",
+			a:     -2,
+			b:     -2,
+			extra: []float64{-2, -2, -2},
+			want:  -10,
+		},
+		{
+			name:  "Sum of 4 numbers to give 0 sum",
+			a:     -1,
+			b:     -1,
+			extra: []float64{1, 1},
+			want:  0,
+		},
+	}
+	for _, tc := range testCases {
+		got := calculator.Add(tc.a, tc.b, tc.extra...)
+		if tc.want != got {
+			t.Errorf("%s: Add(%v, %v, %+v) want %f, got %f",
+				tc.name, tc.a, tc.b, tc.extra, tc.want, got)
+		}
 	}
 }
 
-// func TestSubtract(t *testing.T) {
-// 	t.Parallel()
-// 	var want float64 = 2
-// 	got := calculator.Subtract(4, 2)
-// 	if want != got {
-// 		t.Errorf("want %f, got %f", want, got)
-// 	}
-// }
+func TestSubtract(t *testing.T) {
+	t.Parallel()
+	testCases := []struct {
+		name  string
+		a, b  float64
+		extra []float64
+		want  float64
+	}{
+		{
+			name:  "Difference of 2 equal positive numbers to give 0",
+			a:     1,
+			b:     1,
+			extra: nil,
+			want:  0,
+		},
+		{
+			name:  "Difference of 2 equal negative numbers to give 0",
+			a:     -1,
+			b:     -1,
+			extra: nil,
+			want:  0,
+		},
+		{
+			name:  "Difference of more than 2 arguments",
+			a:     10,
+			b:     1,
+			extra: []float64{1, 1, 1, 1, 1, 1, 1, 1, 1},
+			want:  0,
+		},
+	}
+	for _, tc := range testCases {
+		got := calculator.Subtract(tc.a, tc.b, tc.extra...)
+		if tc.want != got {
+			t.Errorf("%s: Subtract(%v, %v, %+v)) want %f, got %f",
+				tc.name, tc.a, tc.b, tc.extra, tc.want, got)
+		}
+	}
+}
+
+func TestMultiply(t *testing.T) {
+	t.Parallel()
+	rand.Seed(time.Now().UnixNano())
+	numTestCases := 100
+	maxInputValue := 10
+	maxNumExtraInputsPerTestCase := 10
+	type testCase struct {
+		a, b  float64
+		extra []float64
+		want  float64
+	}
+	randomTestCases := make(chan testCase)
+	go func() {
```

Comment:
But there's no real concurrency here, is there? The channel is unbuffered, so the main goroutine must wait until the generator has produced a test case, at which point it runs while the generator goroutine waits to be able to write another value to the channel. Only one goroutine is ever running at once, so it would actually be simpler to do this sequentially instead!

Reply:
Something about using that generator pattern with a channel and goroutine felt a little off. Refactoring the test function to run sequentially helped make the code way more understandable.